### ISSUE TITLE
Allow attributes to be defined on a list item in an ordered or unordered list

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#2716]
Introduce a syntax for defining attributes on a list item in an ordered or unordered (item) list.

Placing a block attribute line above an item list defines attributes for that list. The line also implicitly starts a new list. For example:

```
[#id-of-list-1]
* list one
[#id-of-list-2]
* list two
```

However, there are situations when you need to set attributes on individual items in the list, not the list itself. We need to come up with a syntax / designated location for specifying a block attribute list that applies to a specific list item.

My proposal is to use the first line of a list item as an optional block attribute line for the list item.

```
[#id-of-list]
* [#id-of-list-item-1]
  list item one
* [#id-of-list-item-2]
  list item two
```

(the indentation of the list text is optional)

The rule would be that if the first line of the list item text begins with `[` and ends with `]`, parse that line as a block attribute line and associate the attributes with the list item. We'd only allow a single block attribute line.

This would also work for anchors (which are really just an alternate block attribute line syntax).

```
[[id-of-list]]
* [[id-of-list-item-1]]
  list item one
* [[id-of-list-item-2,optional reference text]]
  list item two
```

This issue replaces #2124 and provides a means to solve #2571.

If this idea works out, we may also consider the same thing for description list terms and table cells.

**Comment by mojavelinux** [comment]
Doing this for a description list term is trickier because it has to fit on a single line. But since we have access to the beginning of the line, we could do:

```
[#id-for-term]term :: description
```

We might also consider putting it after the delimiter, which would be more more consistent with the syntax proposed for an item list.

```
term :: [#id-for-term]
description
```

I like the second proposal best, but it does mean that we couldn't also define attributes for the definition (as they would get promoted to the term).

**Comment by mojavelinux** [comment]
As part of this, an ID/anchor defined this way would be associated directly with the list item and not left behind in the text. In other words, it would be promoted from an inline anchor to a block attribute.

**Comment by elextr** [comment]
So the rule is:

- If the attribute list is before the list item it applies to the list (and starts a new list since it makes no sense in the middle of the list), 
- if it follows the list item markup (* . term:: etc) but has no content on the same line it applies to the item, and 
- if it has any content on the same line it applies to the content

```
[list attributes]
* content

* [list item attributes]
content

* [content attributes] content

* [list item attributes]
[content attributes] content
```

Thats ok, and can apply to description lists too, until somebody pushes to the limit and list item attributes or content attributes takes more than one line and then it will be ambiguous again.  It won't be common, but if they _can_ they _will_ :grin:.

So maybe ambiguous situations would require the writer to use a list continuation to separate list item attributes and content attributes:

```
* [list item attributes]
[more list item attributes]
+
[long content attributes]
[more content attributes]
content
```

That should "just work" unless the processor rejects list continuation before any actual content has been encountered, but that should be easy to change.

**Comment by mojavelinux** [comment]
What are content attributes?

**Comment by elextr** [comment]
My lazy shorthand for attributes for the list item contents.

**Comment by mojavelinux** [comment]
My point is this is meaningless:

```
[content attributes] content
```

There is no such thing in AsciiDoc. (The closest thing is `[content attributes]#content#`, but that isn't relevant here).

We are only dealing with two applications:

```
[list attributes]
* [list item attributes]
list item content
```

The existing patterns for AsciiDoc hold up well to this extrapolation. Block attributes always go on their own line. The only thing we're adding is that the block attributes for the list item are **prefixed** with the list marker. Otherwise, everything stays the same.

In the very off chance that an existing document has a list item whose first line matches a block attribute list, and there are no other lines of text in the list item, that first line will get processed as attributes. But I've never seen that case.

**Comment by mojavelinux** [comment]
To address something you pointed out, we won't allow multiple lines of block attributes on a list item. In general, I don't like multiple block attribute lines anyway because it can get ambiguous. But we can certainly make it a rule in this instance since this is new syntax.

**Comment by elextr** [comment]
> There is no such thing in AsciiDoc. (The closest thing is [content attributes]#content#, but that isn't relevant here).

That was actually what first triggered my comment, but yes its not relevant to the situation of the attribute list being on a separate line.

> We're only dealing with two applications:

But if the writer wants to apply attributes (eg a role) to the first paragraph inside the list item there are three:

```
[#list_id]
* [.list_item_role]
[.para_1_role]
blah blah paragraph with para_1_role
+
and a paragraph in the list item but without para_1_role
```

> In general, I don't like multiple block attribute lines anyway because it can get ambiguous.

I sort of agree, but the original specification in Asciidoc Python was that attribute lists had to be one line, so if it got too long it had to be broken into multiple lists which were combined.  This is documented.  

AFAICT Asciidoctor docs are silent on the subject of attribute lists longer than one line, so I assumed it did the same as Asciidoc Python.

I only sort of agree that multiple lists are bad because if attribute lists are allowed to be longer than a line then a missing ] will consume all the rest of the file as attribute list, which tends to be annoying as it generates many errors as well as the "missing ]" one.  But if the writer had to be explicit about extending the attribute list to many lines in some way it becomes less of a problem for example backslashing the eol the same as multi-line attribute values are extended.

> In the very off chance that an existing document has a list item whose first line matches a block attribute list, and there are no other lines of text in the list item, that first line will get processed as attributes.

The writer just needs to escape the opening [ then.

> But I've never seen that case.

Thats never an argument, as I said "if its possible, somebody will do it".

> we won't allow multiple lines of block attributes on a list item

Making special cases gets confusing, and as I said above, what if an attribute list is too long for one line?

**Comment by mojavelinux** [comment]
> But if the writer wants to apply attributes (eg a role) to the first paragraph inside the list item there are three:

That's never been possible and I don't think it's relevant here. The primary text of the list item is already an implicit construct (not a paragraph), and it can be addressed with CSS from the list item.

**Comment by elextr** [comment]
And in docbook?

**Comment by mojavelinux** [comment]
> Making special cases gets confusing, and as I said above, what if an attribute list is too long for one line?

That's so unlikely, it's not even worth addressing. A block attribute line can't wrap. If it is wrapped, that's a user error.

I'll give it more thought as to whether we should allow stacked block attribute lines on a list item, but I still strongly recommend against the use of this style. As far as I'm concerned, it's a bad practice.

**Comment by mojavelinux** [comment]
> And in docbook?

This has nothing to do with the converter. The primary text of a list item is not a node and thus has no attributes. The node is the list item itself. The text is a property of a node (and thus distinct form other attached blocks).

**Comment by 997293** [comment]
Another issue I guess may be relevant is that currently it seems like we cannot assign a title to a list item. Maybe by figuring out a way to assign an ID to a list item, we can also assign a title to it by a similar mechanism.

**Comment by mojavelinux** [comment]
Yes, this design would open the door to assigning any block attribute to a list item. However, I've never seen a title on a list item used in practice. Can you give an example?

**Comment by 997293** [comment]
![image](https://user-images.githubusercontent.com/59507996/86848480-150ef180-c07c-11ea-96d0-4d4adaabd1f9.png)

So here is an example, where I have a bunch of list items and I want to assign a title to each item. The question is what would be the best way to assign the titles.

Here, I assign titles to the *paragraph* in each item. I think this is just a workaround because what if we have more than one paragraph in one item? I think the most semantically correct way to assign titles to items would be to assign the title directly to the *item* itself rather than the first *paragraph* in the item.

In my opinion, as long as there is the need to have more than one paragraph in a list item, there is the need to assign a title to the list item itself.

**Comment by mojavelinux** [comment]
Aha, what you're looking for is the itemized description list (ordered or unordered). These are terms rather than titles.

```
.Cartesian methods
[ordered]
evidence::
The first was...

analysis::
The second, ...
```

This has been implemented in both Asciidoctor PDF and Asciidoctor EPUB3 but has not yet made its way to core. That's proposed in #821. Let's continue this discussion there (or you can try them out in Asciidoctor PDF today).

**Comment by oliviercailloux** [comment]
As this been implemented? I’d need it for describing multiple choice questions. Using the AsciidoctorJ API, it seems the brackets and text inside brackets is seen as part of the item text, not as attributes of the list item.

````
= A difficult question

How much is 2 + 2?

. [claim-status = false]
  easy, it’s three
. [claim-status = true]
  four
. [claim-status = undecided]
  it can only be 42
````

. Desired: the parser sees, in the list block, three items, each with a named attribute `claim-status` and its corresponding value, and with text of the items being the part after the brackets.
. Observed: the parser sees, in the list block, three items with no attributes and text including the brackets and everything.

**Comment by mojavelinux** [comment]
No, this is a proposal we are taking to the AsciiDoc Language project. Asciidoctor itself cannot just introduce new syntax in AsciiDoc as that would create divergence. Once ratified and approved, we will add the feature to Asciidoctor.

Note that it is possible to add attributes to the list item in the API, so you could use a tree processor extension to move the information there to emulate this behavior.
